File: Prelims/Problem 4/Solution.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;

struct Student {
    string name;
    int marks, age;
};

int main() {
    // Take input
    int n;
    cin >> n;
    Student students[n];
    for (int i = 0; i < n; i++) {
        cin >> students[i].name >> students[i].marks >> students[i].age;
    }

    // First sort by age
    stable_sort(students, students + n, [](const Student& a, const Student& b) {
        if (a.name.length() != b.name.length()) {
            return a.name.length() < b.name.length();
        }
        if (a.marks != b.marks) {
            return a.marks > b.marks;
        }
        return a.age < b.age;
    });

    // Print the names of the students
    for (Student student : students) {
        cout << student.name << endl;
    }
}

File: Prelims/Problem 4/Solution.py
def solve():
    # Take input
    n = int(input())
    students = []
    for i in range(n):
        inp = input().split()
        name = inp[0]
        marks = int(inp[1])
        age = int(inp[2])
        students.append([name, marks, age])

    # First sort on basis of age
    students.sort(key = lambda x: x[2])

    # Then sort on basis of marks
    students.sort(key = lambda x: -x[1])

    # Finally sort on basis of name length
    students.sort(key = lambda x: len(x[0]))

    # Print the names of the students
    for student in students:
        print(student[0])

if __name__ == '__main__':
    solve()